### PR TITLE
Bug/699 context provider responses without payload

### PR DESCRIPTION
**Pull request by kzangeli**
### Description
Made the broker not crash when an empty string is input to the XML/JSON parsing routines.
"OK" is returned on empty payload. Not sure this is the best approach, as the function shouldn't
be called on empty payload and yet there is no real 'generic' error sending a request without payload ...  
Two short unit tests to test this out as it is **very** difficult to provoke this failure 'from the outside', using a functional test.

### CODE REVIEW

File: src/lib/serviceRoutines/postQueryContext.cpp [hunk 1]
```diff
@@ -173,6 +174,19 @@ std::string postQueryContext
 
   cleanPayload = xmlPayloadClean(out.c_str(), "<queryContextResponse>");
 
+  if ((cleanPayload == NULL) || (cleanPayload[0] == 0))
+  {
+    QueryContextResponse qcrs;
+
+    // FIXME P1: What StatusCode should be used here?
+    qcrs.errorCode.fill(SccReceiverInternalError, "invalid context provider response");
```
Comment (fgalan) [inline src/lib/serviceRoutines/postQueryContext.cpp]: I don't think this should be a 500 Internal Error. In general, fails in the connection with a CPr should be resolved as if there wasn't any CPr registered for the given context element, i.e. 404 Not Context Element found, from the perspective of the client. Of course, from an "internal" perspective it's perfectly ok (as you do some lines below) logging it as warning message in the log.

File: src/lib/serviceRoutines/postUpdateContext.cpp [hunk 2]
```diff
@@ -188,6 +188,19 @@ std::string postUpdateContext
 
     cleanPayload = xmlPayloadClean(out.c_str(), "<updateContextResponse>");
 
+    if ((cleanPayload == NULL) || (cleanPayload[0] == 0))
+    {
+      UpdateContextResponse ucrs;
+
+      // FIXME P1: What StatusCode should be used here?
```
Comment (fgalan) [inline src/lib/serviceRoutines/postUpdateContext.cpp]: I don't think this should be a 500 Internal Error. In general, fails in the connection with a CPr should be resolved as if there wasn't any CPr registered for the given context element, i.e. 404 Not Context Element found, from the perspective of the client. Of course, from an "internal" perspective it's perfectly ok (as you do some lines below) logging it as warning message in the log.

Comment (kzangeli) [inline src/lib/serviceRoutines/postUpdateContext.cpp]: ok, sure.  It is a strange situation. It is a 500, but not in the broker, but in the context provider.
A user might not know about context providers and see (orion + context providers) as a black box.
This reasoning led me to return a 500 (+ the FIXME about which code to use :-))

File: src/lib/xmlParse/xmlRequest.cpp [hunk 3]
```diff
@@ -208,6 +208,11 @@ std::string xmlTreat
   xml_document<>  doc;
   char*           xmlPayload = (char*) content;
 
+  if ((content == NULL) || (*content == 0))
```
Comment (fgalan) [inline src/lib/xmlParse/xmlRequest.cpp]: As discussed by skype, include an explanatory comment here.

File: src/lib/jsonParse/jsonRequest.cpp [hunk 4]
```diff
@@ -171,6 +171,11 @@ std::string jsonTreat
   std::string   res   = "OK";
   JsonRequest*  reqP  = jsonRequestGet(request, ciP->method);
 
+  if ((content == NULL) || (*content == 0))
```
Comment (fgalan) [inline src/lib/jsonParse/jsonRequest.cpp]: As discussed by skype, include an explanatory comment here.

Comment (kzangeli) [inline src/lib/jsonParse/jsonRequest.cpp]: Fixed in 5218a38

File: src/lib/xmlParse/xmlRequest.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/xmlParse/xmlRequest.cpp]: Fixed in 5218a38

File: src/lib/serviceRoutines/postUpdateContext.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/serviceRoutines/postUpdateContext.cpp]: Fixed in 5218a38

File: src/lib/serviceRoutines/postQueryContext.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/serviceRoutines/postQueryContext.cpp]: Fixed in 5218a38

**Comment by fgalan** [comment]
LGTM

**Pull request merged**